### PR TITLE
core: basic UI support for T1

### PR DESCRIPTION
**Pull request by mmilata**
Just enough UI ported for `trezorctl ping ping -b` to work. Implementations of `Button`, `Confirm`, and `Text` were moved under `src/trezor/ui/formfactor/{common,t1,tt}/` (not sure if formfactor is even the correct word here).

Known issues:
- [x] fonts are rendered with 23px high rectangle of background color - fixed in https://github.com/trezor/trezor-firmware/pull/1173
- [x] glyphs seem to be rendered too far apart - seems to be the property of the font, we need to decide whether to keep it or trim the advances
- [x] does not typecheck
- [ ] could use some docstrings

### CODE REVIEW

**Comment by prusnak** [comment]
> (anyone can point me to explanation of advance/bearing?)

https://www.freetype.org/freetype2/docs/glyphs/glyphs-3.html

**Comment by prusnak** [comment]
Rectangles of background colors are added here:

https://github.com/trezor/trezor-firmware/blob/01ac2ec48d984f90feaf65fd127131eaf6c15447/core/embed/extmod/modtrezorui/modtrezorui-display.h#L314-L318

(same for `text_right` and `text_center`); and they indeed have hard-coded Y dimensions.

Honestly, I am not sure we still need this. Will try to remove this and see if the UI checks break).

**Comment by prusnak** [comment]
Fixed the background rectangle issue in https://github.com/trezor/trezor-firmware/pull/1173

**Comment by mmilata** [comment]
Pushed 2nd iteration that (mostly) doesn't touch the class hierarchy and the layout is starting to resemble the mockup, PTAL. CI tests for TT pass, however typing has [some room for improvement](https://github.com/trezor/trezor-firmware/blob/f391908e7fbf665fa8bb2b2d4cf0d92e80910dc0/core/src/trezor/ui/formfactor/__init__.py#L73).

Regarding the font spacing it indeed seems to be property of the font used - we need to decide whether we want to keep or reduce the spaces.

d4

**Comment by tsusanka** [comment]
> Regarding the font spacing it indeed seems to be property of the font used - we need to decide whether we want to keep or reduce the spaces.

As @mmilata sent me privately, these are the differences between the two. I do not see any reasons to go with the latter, the readability of the first is good. So @mmilata so feel free to go ahead and include that in this PR.

font-lessspace font-origspace

I will look at the code shortly.

**Comment by tsusanka** [comment]
Looking into the code: it's great progress! I am thinking about the `formfactor` module. How about a following structure:

- trezor/
  - ui/
    - button/
       - \_\_init\_\_.py  # all common code for T1 and TT goes here so we can keep on using `from ... import button`
       - t1.py  # t1 specific code 
       - tt.py  # tt specific code

I am actually wondering if we could reach the state when `t1.py` and `tt.py` includes only constants and no application code (and then maybe be something other than python file but maybe that's too naive). What do you think?

**Comment by matejcik** [comment]
I think the `formfactor` module is a reasonable thing, but _its_ structure should more closely resemble what @tsusanka is suggesting.

As in:
`trezor/ui/button.py` - contains code common to T1 and TT
```
from .formfactor.button import SOME_CONSTANT`
```
`trezor/ui/formfactor/button/__init__.py` - contains redirections
```
if is_t1:
	from .t1 import *
else:
	from .tt import *
```
`trezor/ui/formfactor/button/t1.py` - contains constants
`trezor/ui/formfactor/button/tt.py` - contains constants

I think that contents of `formfactor/button/__init__.py` should be as generic as possible, so that `formfactor/<any>/__init__.py` are all identical files.

The build process would then ideally move `formfactor/button/tt.py` to `formfactor/button.py`, and this should be a drop-in operation wrt file structure; ideally, `formfactor/*` shouldn't even import anything.

Perhaps this could be inlined so that only `trezor/ui/button/*` things exist, but ISTM separating into `formfactor` is cleaner.

**Comment by mmilata** [comment]
Thanks for the great feedback guys! I've reorganized the formfactor module as proposed by @matejcik and renamed it to `model`. PTAL!

Regarding exclusively having constants in the `model` namespace I think there are couple of disadvantages.
* it will lead to code like `if model.IS_T1: ... else: ...` or `if model.HAS_FEATURE_X: ... else: ...` which we wanted to avoid in the first place
* code being present in firmware for device which never executes it (because it's different model), wasting flash space

Nevertheless, using constants wherever possible is probably still good rule of thumb.

File: core/src/trezor/ui/model/button/__init__.py
Comment (mmilata) [inline core/src/trezor/ui/model/button/__init__.py:13]: I could get away without `TYPE_CHECKING` in the other `__init__.py` files. Here, if removed, mypy visits both `if` branches and complains about importing incompatible types. Both types are classes derived from common base class but I don't know how to tell this to mypy.

**Review by tsusanka — COMMENTED** [review]
Concept ACK. Great job, I am really excited about this :tada:.

File: core/embed/extmod/modtrezorui/modtrezorui-display.h
Comment (matejcik) [inline core/embed/extmod/modtrezorui/modtrezorui-display.h:599]: please also drop the corresponding `/// FONT_SIZE` comment

File: core/src/trezor/ui/button.py
Comment (matejcik) [inline core/src/trezor/ui/button.py:6]: perhaps worth only importing the `render_button` function?

File: core/src/trezor/ui/confirm.py
Comment (matejcik) [inline core/src/trezor/ui/confirm.py:13]: thought (not necessarily for this PR):
the usage `from .model import <name> as model` suggests a different possibility:
```
def import_model_specific(name: str) -> Module:
	model_name = "t1" if is_t1 else "tt"
	widget_name = name.split(".")[-1]
	return __import__("trezor.ui.model.{}.{}".format(widget_name, model_name), None, None, (widget_name,), 0)

model = import_model_specific(__name__)
```

This would avoid the need to have any code at all in `model/*/__init__.py`. It would preclude the thing where the build system replaces the `__init__` with the appropriate module directly, but that might not matter.

what do you all think?

Comment (matejcik) [reply core/src/trezor/ui/confirm.py:13]: another drawback is that this would probably be untypeable...

File: core/src/trezor/ui/model/__init__.py
Comment (matejcik) [inline core/src/trezor/ui/model/__init__.py:8]: let's expose a constant in `modtrezorutils` for this. probably model string? enum?

File: core/src/trezor/ui/model/button/t1.py
Comment (matejcik) [inline core/src/trezor/ui/model/button/t1.py:125]: this doesn't do anything istm?

File: core/src/trezor/ui/model/button/t1.py
Comment (matejcik) [inline core/src/trezor/ui/model/button/t1.py:35]: A general note: these style definitions seem to be at the wrong level of abstraction here.

I'm wondering if `Button` should actually be exposed as a widget the way it is now -- as opposed to an implementation detail of, e.g., the `Confirm` widget. Then we could have wildly different implementations of `Button` for T1 and TT -- which seems to be _what we need_ -- because the button implementation would not be visible from outside the `ui.model.<model>` namespace.

I.e., in `trezor/ui/confirm.py`:
```
model = import_model_specifics(__name__)

model.show_confirm("Are you sure?", CONFIRM_TYPE_OKCANCEL)
```
in `trezor/ui/model/t1/confirm.py`
```
from . import button

button.render_left("OK")
button.render_right("Cancel")
```
in `trezor/ui/model/tt/confirm.py`
```
from . import button

button.render("OK", ButtonStyle.ButtonConfirm, GRID_LEFT)
button.render("Cancel", ButtonStyle.ButtonCancel, GRID_RIGHT)
```

...which would necessitate flipping again? I don't remember the original layout, but I think it was `ui.model.<model>.<widget>`, and I proposed `ui.model.<widget>.<model>`? If that's so, then your original version was perhaps the smarter choice.

---

not saying that this needs to happen in this PR, but it seems to be the right general direction?

**Review by matejcik — COMMENTED** [review]
Code mostly LGTM, but see mainly the comment about abstraction levels.

We _could_ merge this now if we wanted, but I'm not sure whether it's worth it without the more high-level changes.

File: core/src/trezor/ui/model/__init__.py
Comment (prusnak) [reply core/src/trezor/ui/model/__init__.py:8]: There is already such constant => `trezor.utils.MODEL`

**Comment by prusnak** [comment]
High-level suggestion. Let's replace:

``` python
if IS_T1:
    from .t1 import *  # noqa: F401,F403
else:
    from .tt import *  # noqa: F401,F403
```

with

``` python
if utils.MODEL == "1":
    from .t1 import *  # noqa: F401,F403
elif utils.MODEL == "T":
    from .tt import *  # noqa: F401,F403
else
    raise ValueError("Unknown Trezor model")
```

File: core/embed/extmod/modtrezorui/modtrezorui-display.h
Comment (mmilata) [reply core/embed/extmod/modtrezorui/modtrezorui-display.h:599]: Fixed 21c051c91d0c2deb7eddd57be7f31a3d4920da20.

File: core/src/trezor/ui/button.py
Comment (mmilata) [reply core/src/trezor/ui/button.py]: Fixed 2290a921f1c430d3742602b38a78be72fb23ceb1.

File: core/src/trezor/ui/model/__init__.py
Comment (mmilata) [reply core/src/trezor/ui/model/__init__.py]: Fixed 0f45a44393684db76692ee4ea8ea14682ac89216.

File: core/src/trezor/ui/model/button/t1.py
Comment (mmilata) [reply core/src/trezor/ui/model/button/t1.py:125]: Fixed 2ccf3a6fa3958e065e0255371f53a4f86887e0ec. I'd expect flake8 to catch this:thinking:.

File: core/src/trezor/ui/model/button/t1.py
Comment (mmilata) [reply core/src/trezor/ui/model/button/t1.py:35]: You're right, the button handling is ugly and the typing problems are a symptom. Let's do one(+) more iteration(s) and get this right before merging.

File: core/Makefile
Comment (prusnak) [inline core/Makefile:133]:
```suggestion
	$(SCONS) CFLAGS="$(CFLAGS)" PRODUCTION="$(PRODUCTION)" TREZOR_MODEL="$(TREZOR_MODEL)" $(BOOTLOADER_CI_BUILD_DIR)/bootloader.bin
```

**Review by prusnak — CHANGES_REQUESTED** [review]
Please also update `SConscript.bootloader_ci` the same way you updated `SConscript.bootloader`

File: core/Makefile
Comment (prusnak) [reply core/Makefile]: Resolved in b9ece6ff66ecf4d54d12d0cd9b6fa0566483595d

**Review by prusnak — APPROVED** [review]
Haven't checked the `core/src` stuff, but changes to `Makefiles`. `SConscript` files and `gen_font.py`  are OK.

**Comment by prusnak** [comment]
I just merged the build system changes in https://github.com/trezor/trezor-firmware/pull/1233 - you can rebase the pull request on top of the current master. Thanks!

File: core/src/trezor/ui/model/button/t1.py
Comment (mmilata) [reply core/src/trezor/ui/model/button/t1.py:35]: I tried to implement this approach in #1239, PTAL. It's less polished than this PR but hopefully should be useful for comparison (typechecks and passes TT tests though).

**Comment by mmilata** [comment]
To be continued in #1239.

**Pull request closed without merging**
